Fine-tuning/few-shot eval support

awebson opened the pull request:
This script extends `run_eval.py` to support few-shot evaluation, which naturally also works for non-few-shot full fine-tuning (when `num_shots is None`, which is actually the default.) Let me know if I should name this `few_shot_eval.py` or `fine_tune.py`. (The former was my own intended use, but the latter may be used by more people?)

In theory, this script could be further extended to support training T0 from scratch (that is, from T5 LM Adapt), but more work needs to be done for handling mixtures of datasets and prompts. 

I replaced `run_eval.py`'s output JSON with a CSV where each row is an epoch (will add support for eval per x steps soon). I also added optional support for Weights & Biases for loss and metric graphs plus template metadata.

VictorSanh reviewed (changes requested):
thanks for contributing this @awebson , it mostly lgtm!

i left a few comments throughout the script.

- I am tempted to move this script into another folder since a/ it is different than T zero (or anything that we report in the paper) but rather, it is an extension that supports your paper, b/ it is doing training+evaluation, not just evaluation. How about "examples"?
- It would helpful to document this script with a README. Something like a readme at the root of the "examples" folder stating "scripts extending the original t zero bla bla". first section would be "Single task (few-shot) fine-tuning"

Guitaricet commented on evaluation/few_shot_eval.py:
```diff
@@ -0,0 +1,699 @@
+        type=str,
+        default=None,
+        help="Where to store the results CSV and (TODO) optionally the final model."
+    )
```
```suggestion
        required=True,
    )
```

The script fails if you don't provide this argument

Guitaricet commented on evaluation/few_shot_eval.py:
```diff
@@ -0,0 +1,699 @@
+from transformers import (
+    AutoConfig,
+    AutoModelForSeq2SeqLM,
+    AutoTokenizer,
```
We need to add sentencepiece to requirements, because the tokenizer fails with this message 

```
ValueError: Couldn't instantiate the backend tokenizer from one of:
(1) a `tokenizers` library serialization file,
(2) a slow tokenizer instance to convert or
(3) an equivalent slow tokenizer class to instantiate and convert.
You need to have sentencepiece installed to convert a slow tokenizer to a fast one.
```

VictorSanh replied:
+1

Guitaricet commented on evaluation/few_shot_eval.py:
```diff
@@ -0,0 +1,699 @@
+    if args.tokenizer_name:
+        tokenizer = AutoTokenizer.from_pretrained(args.tokenizer_name, use_fast=not args.use_slow_tokenizer)
+    elif args.model_name_or_path:
+        tokenizer = AutoTokenizer.from_pretrained(args.model_name_or_path, use_fast=not args.use_slow_tokenizer)
```
I had this error, because of missing protobuff.

```
ImportError:
T5Converter requires the protobuf library but it was not found in your environment. Checkout the instructions on the
installation page of its repo: https://github.com/protocolbuffers/protobuf/tree/master/python#installation and follow the ones
that match your environment.
```

If you follow the provided link, it does not really contain installation instructions 😢. All they have is "Build the C++ code, or install a binary distribution of protoc." without explanation on how to install protoc. I found an easier way to do it (requires Ubuntu>=20.04).

```
sudo apt-get install protobuf-compiler
pip install protobuff
```

Recommend adding this to README.

VictorSanh replied:
`pip install protobuf` usually fixed the issue for me
+1 documentation

Guitaricet commented on evaluation/few_shot_eval.py:
```diff
@@ -0,0 +1,699 @@
+
+
+    # Metrics
+    metric = load_metric("accuracy")
```
Add `sklearn` to requirements.

```ImportError: To be able to use accuracy, you need to install the following dependencies['sklearn'] using 'pip install sklearn' for instance'```

Guitaricet reviewed (commented):
I noted a couple of minor bugs (all of them are commented in the review), the biggest one being `args.model_name_or_path` instead of `args.config_name` on the line 360.

It took me a bit of extra time to run the script. The main problem was the unspecified requirements (promptsource requirements.txt, sklearn, and protobuff).

It also would be great to have a README with a training command. I used 

```
python evaluation/few_shot_eval.py \
    --dataset_name super_glue \
    --dataset_config_name rte \
    --template_name "must be true" \
    --model_name_or_path bigscience/T0_3B \
    -tb 1 -ns 8 \
    --output_dir ./debug
```

Unfortunately, it OOM-ed on the beefiest GPU I currently have access to (3090), so I have not tested the training and evaluation loop.

VictorSanh commented:
+ in the readme, document how much computational resources are required (as suggested in #8, think that's a good idea)

VictorSanh commented on evaluation/few_shot_eval.py:
```diff
@@ -0,0 +1,699 @@
+import csv
+import math
+
+import datasets
```
hmm... they are already set as requirements... https://github.com/bigscience-workshop/promptsource/blob/4a662713df656d37ee1f3fb8ee7966cb2b8e5a74/requirements.txt#L8

Guitaricet replied:
They are in the requirements, but the `requirements.txt` is not used in `setup.py`. I added a proposal to fix that https://github.com/bigscience-workshop/promptsource/pull/711

VictorSanh replied:
oh yes you are totally right, sorry for some reason i was sure the requirements were called in the setup.py file but i was wrong.

i will accept your PR. note @Guitaricet  that since the requirements points to promptsource v0.1 (the version we used for t zero), this PR won't be reflected, so let's just add this to the requirements here

VictorSanh commented on evaluation/few_shot_eval.py:
```diff
@@ -0,0 +1,699 @@
+                global_steps += 1
+                loss = loss.item()
+                if accelerator.is_main_process:
+                    tqdm.write(f"{epoch = }, {global_steps = }, {loss = }")
```
oh nice, i didn't know about self-document expressions

awebson commented on evaluation/few_shot_eval.py:
```diff
@@ -0,0 +1,699 @@
+        help="Batch size (per device) for the evaluation dataloader. Will be multiplied by the number of answer choices.",
+    )
+    parser.add_argument(
+        "-tb",
```
How about short names for all common arguments, while long names for infrequently used arguments? Please take a glance at the new argparse. Happy to revert if we want to follow the HF convention and stick to long names only.

VictorSanh replied:
hmm that sounds super arbitrary to me, i don't think that it's an hf-specific convention. sticking to my comment

awebson replied:
That's fair. I added short names for all arguments to be consistent. I find it kinda important for the user's quality of life. It's always kinda annoying when a terminal command is 3 lines long and you need to jump into the middle (even with Option + F/B), or using backslash to escape newlines in which case the up/down arrow keys sometimes have weird behaviors.

awebson commented on evaluation/few_shot_eval.py:
```diff
@@ -0,0 +1,699 @@
+        if args.num_shots is not None:
+            max_index = len(raw_train_dataset) - args.num_shots
+            start_index = random.randint(0, max_index)
+            sample_indices = range(start_index, start_index + args.num_shots)
```
I'm aware that some datasets are sorted by label. This was a tradeoff where, e.g., in few-shot eval, I wanted to make it easy to check 16-shot is run on examples 10 - 26, and 32-shot is run on examples 10 - 42, etc. where the "higher shots run" should use the same "lower shots run" examples.

That being said, I agree this tradeoff is now outdated and preempting a label-sorted dataset is more important. Changed to non-contiguous slice now:
```python
sample_indices = random.sample(range(0, len(raw_train_dataset)), k=args.num_shots)
raw_train_dataset = raw_train_dataset.select(sample_indices)
```

VictorSanh replied:
great!

awebson commented on evaluation/few_shot_eval.py:
```diff
@@ -0,0 +1,699 @@
+    )
+
+    # Prepare everything with our `accelerator`.
+    optimizer, train_dataloader, eval_dataloader = accelerator.prepare(optimizer, train_dataloader, eval_dataloader)
```
When we call `model.parallelize()`, letting `accelerate` to do device placement again causes an error. I changed this to a conditional: 
```python
    if args.parallelize:
        num_gpus = torch.cuda.device_count()
        assert num_gpus > 1, "You need at least 2 GPUs to use `model.parallelize()`."
        model.parallelize()
        optimizer, train_dataloader, eval_dataloader = accelerator.prepare(
            optimizer, train_dataloader, eval_dataloader)
    else:
        model, optimizer, train_dataloader, eval_dataloader = accelerator.prepare(
            model, optimizer, train_dataloader, eval_dataloader)
```

awebson commented on evaluation/few_shot_eval.py:
```diff
@@ -0,0 +1,699 @@
+                global_steps += 1
+                loss = loss.item()
+                if accelerator.is_main_process:
+                    tqdm.write(f"{epoch = }, {global_steps = }, {loss = }")
```
Ha it's a Python 3.8 [feature](https://docs.python.org/3/whatsnew/3.8.html#f-strings-support-for-self-documenting-expressions-and-debugging), and you guessed what it means exactly correctly! I changed it to make it compatible below 3.8.

awebson commented on evaluation/few_shot_eval.py:
```diff
@@ -0,0 +1,699 @@
+    # In distributed training, the .from_pretrained methods guarantee that only one local process can concurrently
+    # download model & vocab.
+    if args.config_name:
+        config = AutoConfig.from_pretrained(args.model_name_or_path)
```
Oh good catch! I copied this from [run_eval.py](https://github.com/bigscience-workshop/t-zero/blob/e845b5256a4db59bc126d060ce3a4778733337d4/evaluation/run_eval.py#L275) so should change it there too. cc @VictorSanh

VictorSanh replied:
yup Thomas changed it in his PR

awebson commented on evaluation/few_shot_eval.py:
```diff
@@ -0,0 +1,699 @@
+            padding=padding,
+            max_length=args.max_length,
+            truncation=True,
+            add_special_tokens=False,
```
Thanks! I added this as an argparse argument and in README instead of inline comment here.

awebson commented:
Thank you @VictorSanh and @Guitaricet! I accepted most changes. Please see my replies to your comments.

I also added a readme and updated the requirements.txt in the repo root. Although note that this script is not just “an extension that supports my paper”, for which I have a pretty different implementation. I think of this script as preparing for a v3 of the T0 paper (or as a standalone follow-up paper) with few-shot eval, as well as preparing for the inevitable multitask prompted training + eval suite for the BigScience official model.

Relatedly, FLAN added few-shot eval for their v3, although they only used priming (in-context learning). So it’d be worthwhile to compare. I don’t think there exists any public study that compares priming vs. regular gradient updated few-shot learning for any large models?

VictorSanh commented on evaluation/few_shot_eval.py:
```diff
@@ -0,0 +1,699 @@
+            }
+            input, target = template.apply(ex)
+            ex_answer_choices = template.get_answer_choices_list(ex)
+            assert target in ex_answer_choices
```
fair enough

VictorSanh commented on evaluation/few_shot_eval.py:
```diff
@@ -0,0 +1,699 @@
+                if accelerator.is_main_process:
+                    tqdm.write(f"{epoch = }, {global_steps = }, {loss = }")
+                if args.wandb_proj and accelerator.is_main_process:
+                    wandb.log({"loss": loss}, step=global_steps)
```
yes you are totally right, i messed up. sorry

VictorSanh commented on requirements.txt:
```diff
@@ -1,4 +1,8 @@
 git+git://github.com/bigscience-workshop/promptsource@v0.1.0
 accelerate
 transformers
+datasets
```
```suggestion
datasets
jinja2
```

VictorSanh replied:
cc @Guitaricet

Guitaricet replied:
Looks good. I think the new `requirements.txt` covers all of the dependencies issues I had.

VictorSanh commented:
Thanks for making the changes @awebson! Looks good to me!

> I also added a readme and updated the requirements.txt in the repo root. Although note that this script is not just “an extension that supports my paper”, for which I have a pretty different implementation. I think of this script as preparing for a v3 of the T0 paper (or as a standalone follow-up paper) with few-shot eval, as well as preparing for the inevitable multitask prompted training + eval suite for the BigScience official model.

Right, as you noted, there are no results reported in any paper yet so let's not create any confusion on what is anticipating & preparing vs replicating/reproducing the t-zero paper!

awebson commented on examples/README.md:
````diff
@@ -0,0 +1,46 @@
+    --template_name "does this imply" \
+    --model_name_or_path bigscience/T0_3B \
+    --output_dir ./debug
+```
````
Sounds good. I removed the command but kept the links to their documentations in case anyone wants to tweak it themselves. (I will likely give DeepSpeed another try in February.)

awebson commented:
Thank you! I accepted all changes to the readme and requirements.

VictorSanh commented:
thanks a bunch @awebson and @Guitaricet for your comments!
merging now!
